RavenDB-16273: Improved codegen for the Blittable JSON Document Builder

redknightlois opened the pull request.

redknightlois committed:
RavenDB-16273: Improved codegen for the Blittable JSON Document Builder

redknightlois committed:
RavenDB-16273: Avoid using a fixed statement when an Span would do it.

redknightlois committed:
RavenDB-16273: GetPropertyType cache.

ayende commented on src/Raven.Client/Json/Serialization/NewtonsoftJson/Internal/BlittableJsonConverterBase.cs:
```diff
@@ -240,11 +240,36 @@ private static bool ShouldSimplifyJsonBasedOnType(string typeValue)
             return typeof(IEnumerable).IsAssignableFrom(type.GetGenericTypeDefinition());
         }
 
+        internal class TypeCacheItem : IEquatable<TypeCacheItem>
```
Should this be a struct?

redknightlois replied:
No. If we use structs the actual cost of dealing with them explodes.

ayende commented on src/Raven.Client/Json/Serialization/NewtonsoftJson/Internal/BlittableJsonConverterBase.cs:
```diff
@@ -240,11 +240,36 @@ private static bool ShouldSimplifyJsonBasedOnType(string typeValue)
+            Type type = null;
+            var item = new TypeCacheItem() { Property = propertyName, Type = rootType };
+            if (_propertyTypeCache.TryGetValue(item, out var result))
+                return type;
```
Why are you returning `null` here (the only option for `type` if you _got_ a value? Shouldn't this be `result`?
```suggestion
                return result;
```

redknightlois replied:
Good catch. I actually push that one for Jenkins to run the slow tests. Thats why I havent seen improvements on that one when I run the profiling test.

ayende commented on src/Sparrow/Collections/FastStack.cs:
```diff
@@ -208,10 +217,26 @@ public void Push(T item)
 
+        // Pushes an item to the top of the stack.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ref T PushByRef()
```
What is the use case here, it looks like a pretty strange API, no?

redknightlois replied:
It is. But that strange API allows the JIT to generate a great assembler. The reason why that API is useful is that you use the knowledge those objects have been already created in bulk and you just write it's values. Big differences in speed and CPU use

ayende commented on src/Sparrow/Json/ManualBlittableJsonDocumentBuilder.cs:
```diff
@@ -32,27 +33,24 @@ public sealed class ManualBlittableJsonDocumentBuilder<TWriter> : AbstractBlitta
-                State = ContinuationState.ReadObjectDocument
-            });
+            ref var state = ref _continuationState.PushByRef();
+            state.State = ContinuationState.ReadObjectDocument;
```
Wouldn't it be easier to have the `Push` accept the value, instead of a `ref return`?
A `ref` param here to save copying would have the same impact, no?

redknightlois replied:
Nope. That will force emitting a constructor and a copy.

redknightlois committed:
RavenDB-16273: Avoid lookups when accessing well known properties by name

redknightlois committed:
RavenDB-16273: Remove manual optimizations no longer needed.

The JIT has advanced to the point that some optimization are no longer needed and in fact may present the JIT with a challenging time to optimize the code.

redknightlois committed:
RavenDB-16273: GetMaxByteCount will cause virtual calls

Avoiding the virtual calls is important, therefore we will tradeoff memory and prefer bigger buffers instead. 5 times the buffer is enough to ensure that even all unicode character would enter into the buffer.

redknightlois committed:
RavenDB-16273: Simplified a switch statement to avoid the lookup.

redknightlois committed:
RavenDB-16273: Improving the codegen with ReadOnlySpan<byte>

```static ReadOnlySpan<byte>``` and usages has special treatment at JIT level. We can improve a lot of the performance switching as much as possible into using it and its writable version (```Span<byte>```)

ayende commented on src/Sparrow/Json/LazyStringValue.cs:
```diff
@@ -196,15 +196,19 @@ public bool Equals(string other)
-                _lazyStringTempComparisonBuffer = new byte[Bits.PowerOf2(sizeInBytes)];
+            if (_lazyStringTempComparisonBuffer == null || _lazyStringTempComparisonBuffer.Length < other.Length * 5)
+            {
+                var sizeInBytes = Encodings.Utf8.GetMaxByteCount(other.Length);
```
Why bother with this entirely? Can't we just specify that the size is *5 and be done with it entirely?

redknightlois replied:
Yes, but how much do we grow it afterward. I prefer to have a tighter number, which that calls provide us that.

ppekrol commented:
@redknightlois can you rebase?

redknightlois commented:
@ppekrol No, I will create a new PR with only the safe changes.

ayende commented on src/Sparrow/Json/BlittableJsonTextWriter.cs:
```diff
@@ -94,52 +94,62 @@ public abstract unsafe class AbstractBlittableJsonTextWriter : IDisposable
         private const byte Colon = (byte)':';
-        public static readonly byte[] NaNBuffer = { (byte)'"', (byte)'N', (byte)'a', (byte)'N', (byte)'"' };
-        public static readonly byte[] PositiveInfinityBuffer =
+        public static ReadOnlySpan<byte> NaNBuffer => new byte[] { (byte)'"', (byte)'N', (byte)'a', (byte)'N', (byte)'"' };
+        public static ReadOnlySpan<byte> PositiveInfinityBuffer => new byte[]
```
This create _properties_,  which allocates a new value each time.
Pretty sure that isn't what you intended to do.

redknightlois replied:
Nope, those are subjected to a special optimization at the JIT level. What happens there is that those values are baked in at the dll .text section. It is a neat trick :D
https://vcsjones.dev/csharp-readonly-span-bytes-static/

ayende replied:
I remember reading that, but I though it was for span fields, not for properties

ayende commented on src/Sparrow/Json/BlittableJsonTextWriter.cs:
```diff
@@ -94,52 +94,62 @@ public abstract unsafe class AbstractBlittableJsonTextWriter : IDisposable
-        public static readonly byte[] NegativeInfinityBuffer =
+        };        
+
+        public static ReadOnlySpan<byte> NegativeInfinityBuffer => new byte[]
```
Property instead of static field here.

ayende commented on src/Sparrow/Json/BlittableJsonTextWriter.cs:
```diff
@@ -94,52 +94,62 @@ public abstract unsafe class AbstractBlittableJsonTextWriter : IDisposable
 
-        private static readonly byte[] EscapeCharacters;
-        public static readonly byte[][] ControlCodeEscapes;
+        private static ReadOnlySpan<byte> EscapeCharacters => InitializeEscapeCharacters();
+
+        public static ReadOnlySpan<int> ControlCodeEscapes => InitializeControlCodeEscapes();
```
Properties, computed each time

arekpalinski commented on src/Raven.Client/Json/Serialization/NewtonsoftJson/Internal/BlittableJsonWriter.cs:
```diff
@@ -25,6 +30,10 @@ internal class BlittableJsonWriter : JsonWriter, IJsonWriter
+
+            _metadataKey = context.GetLazyStringForFieldWithCaching(Constants.Documents.Metadata.Key);
+            _metadataCollection = context.GetLazyStringForFieldWithCaching(Constants.Documents.Metadata.Collection);
+            _metadataId = context.GetLazyStringForFieldWithCaching(Constants.Documents.Metadata.Id);
```
Shouldn't we dispose them on writer dispose?

ayende replied:
No, note that this is used with caching, so it will live for the life time of the context.
